source/tests/system/visa_driver_api_tests.cpp

```diff
@@ -0,0 +1,192 @@
+#include "device_server.h"
+#include "tests/utilities/test_helpers.h"
+#include "visa/visa_client.h"
+
+namespace ni {
+namespace tests {
+namespace system {
+
+namespace visa = visa_grpc;
+
+class VisaDriverApiTest : public ::testing::Test {
+ protected:
+  VisaDriverApiTest()
+      : device_server_(DeviceServerInterface::Singleton()),
+        visa_stub_(visa::Visa::NewStub(device_server_->InProcessChannel()))
+  {
+    device_server_->ResetServer();
+  }
+
+  virtual ~VisaDriverApiTest() {}
+
+  void SetUp() override
+  {
+#ifndef WIN32
+    GTEST_SKIP() << "Digital pattern is not supported on Linux.";
+#endif
+    initialize_driver_session();
+  }
+
+  void TearDown() override
+  {
+    close_driver_session();
+  }
+
+  std::unique_ptr<visa::Visa::Stub>& GetStub()
+  {
+    return visa_stub_;
+  }
+
+  std::string GetSessionName()
+  {
+    return driver_session_->name();
+  }
+
+  void initialize_driver_session()
+  {
+    ::grpc::ClientContext context;
+    visa::OpenRequest request;
+    visa::OpenResponse response;
+    request.set_instrument_descriptor("TCPIP::www.ni.com::80::SOCKET");
+    request.set_access_mode(visa::LOCK_STATE_NO_LOCK);
+    request.set_session_name("SessionName");
+    request.set_open_timeout(VI_TMO_IMMEDIATE);
+    ::grpc::Status status = GetStub()->Open(&context, request, &response);
+    nidevice_grpc::experimental::client::raise_if_error(status, context);
+    driver_session_ = std::make_unique<nidevice_grpc::Session>(response.vi());
+
+    EXPECT_TRUE(status.ok());
+    EXPECT_EQ(VI_SUCCESS, response.status());
+  }
+
+  void close_driver_session()
+  {
+    if (!driver_session_) {
+      return;
+    }
+    ::grpc::ClientContext context;
+    visa::CloseRequest request;
+    request.mutable_vi()->set_name(driver_session_->name());
+    visa::CloseResponse response;
+    ::grpc::Status status = GetStub()->Close(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    EXPECT_EQ(VI_SUCCESS, response.status());
+  }
+
+  void set_bool_attribute(visa::VisaAttribute attribute, bool value)
+  {
+    ::grpc::ClientContext context;
```

Comment:
I don't have a problem with writing the driver API tests like this, but it's usually much easier to write them with the generated clients. All of the services added since daq (I think) use those for the driver API.

Reply:
Yep, the client code does indeed look nicer. I modified most VISA driver tests to take advantage of it.